### PR TITLE
Add initial Address allocator implementation

### PR DESCRIPTION
**Pull request by joergroedel**
The address allocator uses a first-fit algorithm and an RBTree as its data structure. This is just a start, more features will be needed in the future, but here is the initial version for review.

### CODE REVIEW

**Review by 00xc — CHANGES_REQUESTED** [review]
As per Carlos Bilbao's RFC on the mailing list, `pub` structs and methods should have some documentation.

File: src/mm/address_ranges.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            data,
+        }
+    }
+
+    fn get(&self) -> &T {
+        &self.data
+    }
+}
+
+intrusive_adapter!(RangeAdapater<T> = Box<Range<T>>: Range::<T> { link: Link });
```
Comment (00xc) [inline src/mm/address_ranges.rs:38]: Small typo, should be `RangeAdapter`

File: src/mm/address_ranges.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        let mut start = align_up(self.pfn_low, align);
+        let mut end = start;
+
+        if size == 0 {
+            return None;
+        }
+
+        let mut cursor = self.tree.front_mut();
+        while !cursor.is_null() {
+            let node = cursor.get().unwrap();
```
Comment (00xc) [inline src/mm/address_ranges.rs:76]: This is equivalent and avoids an unwrap:

```rust
while let Some(node) = cursor.get() {
    ...
}
```

File: src/mm/address_ranges.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        if end > start && end - start >= size {
+            cursor.insert_before(Box::new(Range::<T>::new(start, start + size, data)));
+            Some(start << PAGE_SHIFT)
+        } else {
+            None
+        }
+    }
+
+    pub fn alloc(&mut self, size: usize, data: T) -> Option<usize> {
```
Comment (00xc) [inline src/mm/address_ranges.rs:97]: Maybe the return value should be a `VirtAddr`? Same with the input for `free()` and `get()`

File: src/mm/address_ranges.rs (hunk 1 above)
Comment (joergroedel) [reply src/mm/address_ranges.rs:38]: Fixed, thanks.

File: src/mm/address_ranges.rs (hunk 2 above)
Comment (joergroedel) [reply src/mm/address_ranges.rs:76]: Yes, this is better, thanks. Changed it.

File: src/mm/address_ranges.rs (hunk 3 above)
Comment (joergroedel) [reply src/mm/address_ranges.rs:97]: Converted to VirtAddr.

**Comment by joergroedel** [comment]
During further development I realized that a separate generic address allocator is not the best way, because it will cause a lot of issues with locking. Going forward it will be integrated into the data structure that also keeps track of virtual memory mappings and page-tables. So I close this PR for now, the code here will be submitted as part of another PR.

**Pull request closed without merging**